bedrock/cms/tests/test_utils.py

```diff
@@ -0,0 +1,144 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+
+import pytest
+from wagtail.coreutils import get_dummy_request
+from wagtail.models import Locale, Page
+
+from bedrock.cms.utils import get_cms_locales_for_path, get_locales_for_cms_page, get_page_for_request
+
+pytestmark = [pytest.mark.django_db]
+
+
+def test_get_locales_for_cms_page(tiny_localized_site):
+    en_us_homepage = Page.objects.get(locale__language_code="en-US", slug="home")
+    en_us_test_page = en_us_homepage.get_children()[0]
+
+    # By default there are no aliases in the system, so all _locales_available_for_cms will
+    # match the pages set up in the tiny_localized_site fixture
+    assert Page.objects.filter(alias_of__isnull=False).count() == 0
+
+    assert sorted(get_locales_for_cms_page(en_us_test_page)) == ["en-US", "fr", "pt-BR"]
+
+    # now make aliases of the test_page into Dutch and Spanish
+    nl_locale = Locale.objects.create(language_code="nl")
+    es_es_locale = Locale.objects.create(language_code="es-ES")
+
+    nl_page_alias = en_us_test_page.copy_for_translation(locale=nl_locale, copy_parents=True, alias=True)
+    nl_page_alias.save()
+
+    es_es_page_alias = en_us_test_page.copy_for_translation(locale=es_es_locale, copy_parents=True, alias=True)
+    es_es_page_alias.save()
+
+    assert Page.objects.filter(alias_of__isnull=False).count() == 4  # 2 child + 2 parent pages, which had to be copied too
+
+    # Show that the aliases don't appear in the available locales
+    assert sorted(get_locales_for_cms_page(en_us_test_page)) == ["en-US", "fr", "pt-BR"]
+
+
+def test_get_locales_for_cms_page__ensure_draft_pages_are_excluded(tiny_localized_site):
+    en_us_homepage = Page.objects.get(locale__language_code="en-US", slug="home")
+    en_us_test_page = en_us_homepage.get_children()[0]
+    fr_homepage = Page.objects.get(locale__language_code="fr", slug="home-fr")
+    fr_test_page = fr_homepage.get_children()[0]
+
+    fr_test_page.unpublish()
+
+    assert sorted(get_locales_for_cms_page(en_us_test_page)) == ["en-US", "pt-BR"]
+
+
+@pytest.mark.parametrize(
+    "path, expected_page_url",
+    [
+        (
+            "/en-US/test-page/",
+            "/en-US/test-page/",
+        ),
+        (
+            "/test-page/",
+            "/en-US/test-page/",
+        ),
+        (
+            "/pt-BR/test-page/",
+            "/pt-BR/test-page/",
+        ),
+        (
+            "/pt-BR/test-page/child-page/",
+            "/pt-BR/test-page/child-page/",
+        ),
+        (
+            "/fr/test-page/",
+            "/fr/test-page/",
+        ),
+        (
+            "/fr/test-page/child-page/",
+            "/fr/test-page/child-page/",
+        ),
+        # These two routes do not work, even though a manual test with similar ones
```

Comment:
**note:** I poured a lot of time into trying to work out why the page tree did contain this `grandchild-page`, when it's clear it's there from the fixture and if you ask for it's `url_path` in the debugger.  Talking with core Wagtail people too, didn't seem to reveal what was up. Tested manually, though, so I'm OK that this is a false negative. Will be nice to work out why, though. Hopefully it's not a blocker to an r+